BIP138: Compact Encryption Scheme for Non-seed Wallet Data (#1951)
bitcoin/bips: bip-encrypted-backup.md

```
BIP: ?  
Layer: Applications  
Title: Compact encryption scheme for non-seed wallet data  
Author: // TBD  
Comments-Summary: No comments yet.  
Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-????  
Status: Draft  
Type: Informational  
Created: 2025-08-22  
License: BSD-2-Clause  
```

## Introduction

### Abstract

This BIP defines a compact encryption scheme for **wallet descriptors** (BIP-0380),
**wallet policies** (BIP-0388), **labels** (BIP-0329), and

> **wallet backup metadata** (json). The payload must not contains any private key material.

Author: done

```
BIP: ?  
Layer: Applications  
Title: Bitcoin Encrypted Backup  
Author: // TBD  
Comments-Summary: No comments yet.  
Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-????  
Status: Draft  
Type: Informational  
Created: 2025-08-22  
License: BSD-2-Clause  
```

## Introduction

### Abstract

This BIP defines a compact encryption scheme for **wallet descriptors** (BIP-0380),
**wallet policies** (BIP-0388), **labels** (BIP-0329), and
**wallet backup metadata** (json). The encrypted output—called a

> *bitcoin encrypted backup* (BEB), enables users to outsource long‑term storage to

Reviewer: This scheme is generic enough that I don't think it needs the word "bitcoin" in it. I also don't think we need to pick a "marketing name", something more boring like "BIP_XXX encrypted backup" should be fine.

miniscript contexts, keys alone are **not sufficient** for recovery: without the
original descriptor the wallet cannot reconstruct the script.

Offline storage of descriptors has two practical obstacles:

1. **Descriptors are hard to store offline.**  
   Descriptor string representation can be far longer than a 12/24-word seed phrase.  
Paper, steel, and other long-term analog media quickly become impractical for such
lengths, or error-prone to transcribe.

2. **Online redundancy carries risk.**  
   Keeping backups on USB thumb-drives, computers, phones, or (worst) cloud drives
avoids the first problem but amplifies surveillance risk: anyone who gains these
**plaintext descriptors** learns the wallet’s public keys, script structure,
etc...  Even with encryption at the cloud provider, an attacker or a subpoena can
compel access, and each extra copy multiplies the attack surface.

These constraints lead to an acute need for an **encrypted**, and
ideally compact backup format that:

* can be **safely stored in multiple places**, including untrusted on-line services,  
* can be **decrypt only by intended holders** of specified public keys,  

> * and keeps both **keys and plaintext** hidden from any party who lacks the

Author: I'll drop this one, it's redunndant with the previous one

for example, cloud providers, specialized services, etc.  
* **Has access control**: decrypting it should only be available to the desired
parties (typically, a subset of the cosigners).  
* **Easy to implement**: it should not require any sophisticated tools.  
* **Vendor-independent**: it should be easy to implement using any hardware signing
device.  

### Scope

This specification applies to creating, encrypting and decrypting a *Bitcoin Encrypted Backup*
of **public-key-only wallet material**, namely:

- a **descriptor** (BIP-0380)  
- a **wallet policy** (BIP-0388)  
- **BIP-329 labels** (JSONL)  
- **wallet-wide JSON backup metadata**  

All four payloads are encrypted under the same key derivation rules detailed below.
The scheme is **not** intended for descriptors or policies that contain private keys,
seeds, mnemonics, or any other secret key material, nor is it designed for real-time,
interactive workflows.

## Specification

Note: in the followings sections, the operator ⊕ refers to the bitwise XOR.

### Secret generation

* Let `p_1`, `p_2`, .., `p_n` be the public keys in the descriptor/wallet policy, in
increasing lexicographical order.  

> * let `secret` = sha256("BEB_DECRYPTION_SECRET" ‖ `p1` ‖ `p2` ‖ ... ‖ `pn`)  

Reviewer: Liana could switch to the BIP number once assigned and when reading also support BEP. That also gives you the flexibility to make breaking changes to other parts of this spec without having to bump the version number.

> * let `secret` = sha256("BEB_DECRYPTION_SECRET" ‖ `p1` ‖ `p2` ‖ ... ‖ `pn`)  

Author: Agree, we are in control of what happen in Liana, but my concern is more about other implementations that will not be able to decrypt early version generated w/ Liana...
Btw, we already plan to ship w/ version 0, that's why this BIP draft define version 1...

> * let `si` = sha256("BEB_INDIVIDUAL_SECRET" ‖ `pi`)  

Reviewer: If it's possible to use subscript in markdown then I would prefer to keep the original notation with small i. Otherwise let's use an underscore, i.e. `s_i`

Author: done

0x42`).

#### Version

`VERSION`: 1 byte unsigned integer representing the format version. The current
specification defines version `0x01`.

#### Derivation Paths

 Note: the derivation-path vector should not contain duplicates.  
 Derivation paths are optional; they can be useful to simplify the recovery process
if one has used a non-common derivation path to derive his key.

`DERIVATION_PATH` follows this format:

`COUNT`  
`CHILD_COUNT` `CHILD` `...` `CHILD`  
`...`  
`CHILD_COUNT` `CHILD` `...` `CHILD`  

`COUNT`: 1-byte unsigned integer (0–255) indicating how many derivation paths are
included.  
`CHILD_COUNT`: 1-byte unsigned integer (1–255) indicating how many children are in
the current path.  
`CHILD`: 4-byte big-endian unsigned integer representing a child index per BIP-32.

#### Individual Secrets

At least one individual secret must be supplied.

The `INDIVIDUAL_SECRETS` section follows this format:

`COUNT`  
`INDIVIDUAL_SECRET`  
`INDIVIDUAL_SECRET`

`COUNT`: 1-byte unsigned integer (1–255) indicating how many secrets are included.  
`INDIVIDUAL_SECRET`: 32-byte serialization of the derived individual secret.

#### Content

> `CONTENT`: 1-byte unsigned integer identifying what has been encrypted.

Reviewer: I would prefer to keep this secret _and_ have multiple pieces of content in a single backup. So my suggestion is to move this inside the encrypted payload and then allow for multiple content entries.

Author: Yeah this BIP draft differ a bit from the initial post in the fact it intentionally add some formating convention

| Value  | Definition                             |
|:-------|:---------------------------------------|
| 0x00   | Undefined                              |
| 0x01   | BIP-0380 Descriptor (string)           |
| 0x02   | BIP-0388 Wallet policy (string)        |
| 0x03   | BIP-0329 Labels (JSONL)                |

> | 0x04   | Wallet backup (JSON)                   |

Reviewer: Maybe point out that unlisted types should still be decrypted and that new types may be added to the BIP.

That said, a more generic approach would be to allow a bit more space for the type. E.g. a ~4 digit number that must represent a BIP number. Or a short string like "BIP-0329". Both avoid the need to maintain a central register.

Also, if you allow multiple pieces of content in a single encrypted blob, it's useful to have a padding type.

| 0x00   | Undefined                              |
| 0x01   | AES-GCM-256                            |

`NONCE`: 12-byte nonce for AES-GCM-256.  
`LENGTH`: [compact
size](https://en.bitcoin.it/wiki/Protocol_documentation#Variable_length_integer)
integer representing ciphertext length.  
`CIPHERTEXT`: variable-length ciphertext.

Note: `CIPHERTEXT` is followed by the end of the `ENCRYPTED_PAYLOAD` section.  
Compliant parsers MUST stop reading after consuming `LENGTH` bytes of ciphertext;  
additional trailing bytes are reserved for vendor-specific extensions and MUST
be ignored.

## Rationale

 - Why derivation paths are optional: When standard derivation paths are used, they are
 easily discoverable, making them straightforward to brute-force. Omitting them
 enhances privacy by reducing the information shared publicly about the descriptor
 scheme.

- Why avoid including fingerprints in plaintext encoding: Including fingerprints leaks
direct information about the descriptor participants, which compromises privacy.

### Future Extensions

The version field enables possible future enhancements:

- Additional encryption algorithms  

> - Support for threshold-based decryption

Reviewer: - hide the number of participants
- bech32m export (convenient for text based storage like an email to yourself, as well as for printing - if you're patient)

Author: > hide the number of participants

hum, the number of participant can also be hidden (derivation path are optionals) and one can also add fake ones, how do you think we can enhance this further?

Reviewer: It's currently revealed by the number of s_i shares, so it would be a more complicated overhaul. I don't think it's urgent.

Author: added

and backup metadata.  

### Copyright

This BIP is licensed under the BSD 2-Clause License.  
Redistribution and use in source and binary forms, with or without modification, are
permitted provided that the above copyright notice and this permission notice appear
in all copies.

### Motivation

Losing the **wallet descriptor** (or **wallet policy**) is just as catastrophic as
losing the seed itself. The seed lets you sign, but the descriptor maps you to your coins.
For multisig or miniscript wallets, keys alone won't help—without the descriptor, you
can't reconstruct the script.

Offline storage of descriptors has two practical obstacles:

1. **Descriptors are hard to store offline.**  
   Descriptors can be much longer than a 12/24-word seed. Paper and steel backups
   become impractical or error-prone.  

2. **Online redundancy carries privacy risk.**  
   USB drives, phones, and cloud storage solve the length problem but expose your
   wallet structure. Plaintext descriptors leak your pubkeys and script details.
   Cloud encryption doesn't help against subpoenas or provider breaches, and each
   copy increases attack surface.  

> These constraints lead to an acute need for an **encrypted**, and

Reviewer: "acute"? It's been years :-)

> This BIP therefore proposes an **encrypted**, and **compact** backup format that:

```

> BIP: ?

Reviewer 2: ```suggestion
BIP: ?
Layer: Applications
```

Title: Compact encryption scheme for non-seed wallet data

> Author: Pyth <pythcoiner@wizardsardine.com>

Reviewer 3: ```suggestion
Authors: Pyth <pythcoiner@wizardsardine.com>
```

> Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-xxxx

Reviewer 2: The Comments-URI header is no longer used.
```suggestion
```

Status: Draft
Type: Specification

> Created: 2025-08-22

Reviewer 3: ```suggestion
Assigned: ?
```

License: BSD-2-Clause

> Post-History: https://delvingbitcoin.org/t/a-simple-backup-scheme-for-wallet-accounts/1607/31

Reviewer 3: ```suggestion
Discussion: https://delvingbitcoin.org/t/a-simple-backup-scheme-for-wallet-accounts/1607/31
Requires: <BIP number(s)>
```

>               https://groups.google.com/g/bitcoindev/c/5NgJbpVDgEc

Reviewer 3: nit, link to the top post
```suggestion
              https://groups.google.com/g/bitcoindev/c/5NgJbpVDgEc/m/TtGK9sF9BgAJ
```

Redistribution and use in source and binary forms, with or without modification, are
permitted provided that the above copyright notice and this permission notice appear
in all copies.

### Motivation

Losing the **wallet descriptor** (or **wallet policy**) is just as catastrophic as
losing the seed itself. The seed lets you sign, but the descriptor maps you to your coins.
For multisig or miniscript wallets, keys alone won't help—without the descriptor, you
can't reconstruct the script.

Offline storage of descriptors has two practical obstacles:

1. **Descriptors are hard to store offline.**  
   Descriptors can be much longer than a 12/24-word seed. Paper and steel backups
   become impractical or error-prone.

2. **Online redundancy carries privacy risk.**  
   USB drives, phones, and cloud storage solve the length problem but expose your
   wallet structure. Plaintext descriptors leak your pubkeys and script details.
   Cloud storage is often unencrypted, and even cloud encryption could be compromised,
   depending on (often opaque) implementation details. Its security also reduces to
   that of the weakest device with cloud access. Each copy increases the attack surface.

> This BIP therefore proposes an **encrypted**, and compact backup format that:

Reviewer 2: Stumbled on the "and" here:

```suggestion
This BIP therefore proposes an **encrypted**, compact backup format that:
```

> * can be **safely stored in multiple places**, including untrusted on-line services,  

Reviewer 2: ```suggestion
* can be **safely stored in multiple places**, including untrusted online services,  
```

    X-only public keys are 32 bytes, a natural size for cryptographic operations.
    This format is also used in BIP340 (Schnorr signatures) and BIP341 (Taproot).

The normalization process depends on the key type:

#### Extended Public Keys (xpubs)

For extended public keys (including those with origin information and/or multipaths):
- Extract the root extended public key
- Extract the **x-coordinate** from its public key (32 bytes)
- Ignore derivation paths, origin information, and multipath specifiers

#### Compressed Public Keys

For 33-byte compressed public keys (0x02 or 0x03 prefix):
- Remove the prefix byte
- Result is 32 bytes (x-coordinate only)

#### X-only Public Keys

Already in the correct format—use as-is (32 bytes).

#### Uncompressed Public Keys

For 65-byte uncompressed public keys (0x04 prefix):
- Extract the x-coordinate (bytes 1-32)
- Result is 32 bytes

See [keys_types.json](./bip-encrypted-backup/test_vectors/keys_types.json) for
normalization test vectors.

### Encryption

The format uses CHACHA20_POLY1305 (RFC 8439) as the default encryption algorithm,
with a 96-bit random nonce and a 128-bit authentication tag to provide confidentiality

> and integrity. AES_256_GCM is also supported as an alternative.[^chacha-default]

Reviewer: I'm not sure it's a good idea to support two formats. On the encryption side there's no extra burden, since implementers can just pick their favourite. But on the decryption side you kinda have no choice but to support both.

My preference would be to drop AES_256_GCM entirely. Alternatively to make it a second class citizen, with phrasing along the lines of:

> Implementations MUST support CHACHA20_POLY1305 decryption and MAY also support the alternative AES_256_GCM scheme for decryption . 

> Implementations SHOULD use CHACHA20_POLY1305 encryption, but if implementation is problematic they MAY use AES_256_GCM instead.
